php/src/Communicator.cpp

```diff
@@ -57,22 +55,20 @@ class ActiveCommunicator : public IceUtil::Shared
     int expires;
     std::chrono::steady_clock::time_point lastAccess;
 };
-typedef IceUtil::Handle<ActiveCommunicator> ActiveCommunicatorPtr;
+using ActiveCommunicatorPtr = shared_ptr<ActiveCommunicator>;
 
 class FactoryWrapper;
 using FactoryWrapperPtr = shared_ptr<FactoryWrapper>;
 
 class DefaultValueFactory;
 using DefaultValueFactoryPtr = shared_ptr<DefaultValueFactory>;
 
-//
 // CommunicatorInfoI encapsulates communicator-related information that
 // is specific to a PHP "request". In other words, multiple PHP requests
 // might share the same communicator instance but still need separate
 // workspaces. For example, we don't want the value factories installed
 // by one request to influence the behavior of another request.
-//
-class CommunicatorInfoI : public CommunicatorInfo
+class CommunicatorInfoI final : public CommunicatorInfo, public enable_shared_from_this<CommunicatorInfoI>
```

Comment:
Why do we have virtual functions in this final class?

If 'virtual' means 'override', they should be replaced by 'final'.